### DISCUSSION TITLE
Sweeper deployed on its own - had to create some Dummy classes in the Publisher

### DISCUSSION THREAD

**Discussion by IlSocio** [BrighterCommand/Brighter#2059] [Show and tell]
I'd like to totally delegate to the Sweeper the push of the messages to the servicebus.
The publisher component will only insert messages to the outbox table.
So, ideally, the publisher should only contain the configuration to access the database.

To support this scenario, I've been forced to create the two dummy classes below... for the Publisher.
and I'm wondering if there exists a cleaner way to do it.

```csharp
public class DummyProducerRegistryFactory : IAmAProducerRegistryFactory
{
    private readonly IEnumerable<Publication> _publications;

    public DummyProducerRegistryFactory(
        IEnumerable<Publication> publications)
    {
        _publications = publications;
    }

    public IAmAProducerRegistry Create()
    {
        var producers = new Dictionary<string, IAmAMessageProducer>();
        foreach (var publication in _publications)
        {
            producers[publication.Topic] = new DummyProducer();
        }
        return new ProducerRegistry(producers);
    }
}

public class DummyProducer : IAmAMessageProducerAsync
{
    public int MaxOutStandingMessages { get; set; } = -1;
    public int MaxOutStandingCheckIntervalMilliSeconds { get; set; }

    public void Dispose()
    {
    }

    public Task SendAsync(Message message)
    {
        throw new System.NotImplementedException();
    }
}
```

I use them in the Publisher as below:
```csharp
    .UseExternalBus(
        new DummyProducerRegistryFactory(
            new Publication[] {
                new Publication() { Topic = new RoutingKey(LegacyMessageJsonMapper.Topic) },
                new Publication() { Topic = new RoutingKey(GreetingChangedEventMessageMapper.Topic) },
            }).Create()
    )
```

Some relevant part of the code, which forced me to create the DummyProducer class https://github.com/BrighterCommand/Brighter/blob/2efd443d255e01ac6c49f2611e5698b935ff88a8/src/Paramore.Brighter/ExternalBusServices.cs#L90

**Comment by preardon** [comment]
I run the Sweeper as its own thing, so the thing to not is that it looks for a specific publisher or uses the default (i.e. you have to have 1)

My runs as follows

```csharp
var producerRegistry = new AzureServiceBusProducerRegistryFactory(serviceBusClientProvider,
                new AzureServiceBusPublication[]
                {
                    new() {MakeChannels = OnMissingChannel.Validate, Topic = new RoutingKey("default")}
                }, boxSettings.BatchChunkSize).Create();

builder.Services.AddBrighter()
                .UseExternalBus(producerRegistry)
                .UseMsSqlOutbox(outboxSettings, outboxType)
                .UseOutboxSweeper(options =>
                {
                    options.TimerInterval = boxSettings.OutboxSweeperInterval;
                    options.MinimumMessageAge = boxSettings.MinimumMessageAge;
                    options.BatchSize = boxSettings.BatchSize;
                    options.UseBulk = boxSettings.UseBulk;
                });
    )
```

I hope this helps

**Reply by IlSocio** [reply]
So, you had to provide a proper servicebus connectionstring witin the "serviceBusClientProvider" right?
That's the thing I'd like to avoid.

**Reply by preardon** [reply]
@IlSocio ah my apologies, I thought you were asking how to get the sweeper working and not the other way around.

To be honest you should be able to take the same approach. the main thing here is that you should not use the Post or Clear methods on the commandProcessor.

Give me a little bit I'll see what I can workout for you

**Reply by preardon** [reply]
ok, so at this minute there is no way to have an outbox without a producer, however if you want a solution that will work around about as well as your Dummy producer above, you can just an incorrectly configured producer.  I will however discuss this with the guys to see what we can do.

In my experience you maybe want both the ability to do both, I generally send implicitly (deposit post and let the sweeper take care of dispatching) however in some higher importance scenarios you may want to explicitly clear (deposit and clear) to give a higher priority.

**Reply by IlSocio** [reply]
Thanks for the feedback and the ideas @preardon 
Unfortunately, the current implementation of the AzureServiceBusProducerRegistryFactory requires a valid servicebus connectionstring.
The only alternative I found, was to use a MsSqlProducerRegistryFactory instead of the Dummy one... But then I thought it could create confusion to see the reference to "MsSql"... and at the end, I kept the Dummy to let it state explicitely that it doesn't take an "active" part in the flow, its only purpose is to provide a registry for the available topics.

Edit: oh, sorry, I noticed I misread your message...
You were referring to the producer, not to the factory.
Honestly I can't remember if I tried this.
I'll give it a shot, so, at least, I can get rid of the dummyproducer class.

**Reply by preardon** [reply]
@IlSocio when it says a "valid" one I don't think it checks more than string.isNullOrEmpty

**Reply by IlSocio** [reply]
As soon as the `AzureServiceBusMessageProducer` is instantiated an instance of the Microsoft's `ServiceBusAdministrationClient` gets created, which requires a valid connectionstring :(

This is the order of the calls... 
The `AzureServiceBusMessageProducer` takes a `AdministrationClientWrapper` as dependency, which, as soon as it is instantiated will call the:
`ServiceBusConnectionStringClientProvider.GetServiceBusAdministrationClient()`
which in turns, instantiates the microsoft:
`ServiceBusAdministrationClient(connectionString)`


[AzureServiceBusMessageProducer](https://github.com/BrighterCommand/Brighter/blob/ccae7e387aba05c4bf8c0161a85d1d9efd50a902/src/Paramore.Brighter.MessagingGateway.AzureServiceBus/AzureServiceBusMessageProducer.cs#L41)

[AdministrationClientWrapper.Initialize](https://github.com/BrighterCommand/Brighter/blob/ccae7e387aba05c4bf8c0161a85d1d9efd50a902/src/Paramore.Brighter.MessagingGateway.AzureServiceBus/AzureServiceBusWrappers/AdministrationClientWrapper.cs#L28)
[AdministrationClientWrapper](https://github.com/BrighterCommand/Brighter/blob/master/src/Paramore.Brighter.MessagingGateway.AzureServiceBus/AzureServiceBusWrappers/AdministrationClientWrapper.cs#L175)

[ServiceBusConnectionStringClientProvider](https://github.com/BrighterCommand/Brighter/blob/ccae7e387aba05c4bf8c0161a85d1d9efd50a902/src/Paramore.Brighter.MessagingGateway.AzureServiceBus/ClientProvider/ServiceBusConnectionStringClientProvider.cs#L44)
